std/experimental/safeint.d

```diff
@@ -0,0 +1,1164 @@
+                    }
+                }
+                else static if (!Signed && !isSigned!(typeof(vVal)) &&
+                        typeof(vVal).sizeof > T.sizeof)
+                {
+                    if(greater(vVal, this.max))
+                    {
+                        //this.value = this.nan;
+                        this.store.setNaN();
+                        return;
+                    }
+                }
+                else static if (!Signed && isSigned!(typeof(vVal)))
+                {
+                    if (less(vVal, this.min) || greater(vVal, this.max))
+                    {
+                        //this.value = this.nan;
+                        this.store.setNaN();
+                        return;
+                    }
+                }
+
+                this.store.unsetNaN();
+                this.store.value = cast(T)vVal;
+            }
+        }
+    }
+
+    /** Check if this SafeInts value is NaN.
+
+    Returns:
+        true is value is NaN, false otherwise.
+    */
+    @property bool isNaN() const pure
+    {
+        return this.store.isNaN();
+    }
+
+    private static auto getValue(V)(V vIn) pure
+    {
+        static if (isSafeInt!V)
+        {
+            return vIn.store.value;
+        }
+        else
+        {
+            return vIn;
+        }
+    }
+
+    /** This implements $(D +=, -=, %=, /=, *=) for this SafeInt.
+
+    Returns:
+        a copy of this SafeInt.
+    */
+    typeof(this) opOpAssign(string op, V)(V vIn) pure
+    {
+        enum call = "this = this " ~ op ~ " vIn;";
+        mixin(call);
+        return this;
+    }
+
+    /** This implements $(D +, -, %, /, *) for this SafeInt.
+
+    If the result of the operation can not be stored by the SafeInt!T,
+    the resulting value is nan.
+
+    Returns:
+        a new SafeInt!T with the result of the operation.
+    */
+    Unqual!(typeof(this)) opBinary(string op, V)(V vIn) const pure
+    {
+        auto v = getValue(vIn);
+
+        static if (typeof(v).sizeof > 4 || typeof(this.store.value).sizeof > 4)
+        {
+            alias SignedType = long;
+            alias UnsignedType = ulong;
+        }
+        else
+        {
+            alias SignedType = int;
+            alias UnsignedType = uint;
+        }
+
+        bool overflow = false;
+        bool wasNaN = true;
+
+        if (this.isNaN)
+        {
+            auto tmp = Unqual!(typeof(this))();
+            tmp.store.setNaN();
+            return tmp;
+        }
+        static if (isSafeInt!V)
+        {
+            if(vIn.isNaN)
+            {
+                auto tmp = Unqual!(typeof(this))();
+                tmp.store.setNaN();
+                return tmp;
+            }
+        }
+
+        static if (op == "+")
+        {
+            static if (T.sizeof > 4 || typeof(v).sizeof > 4)
+            {
+                long function(long, long, ref bool)
+                    @safe @nogc pure nothrow sOp = &adds;
```

Comment:
This might be problematic if the ``sOp``/``uOp`` pointers aren't folded in time for the front-end to check whether they are intrinsics or not.  I'd advise on using ``alias sOp = adds`` instead.